File: Programmers/코딩테스트 고득점 Kit/해시/완주하지_못한_선수.py
# 완주하지 못한 선수
# https://programmers.co.kr/learn/courses/30/lessons/42576

"""
    입출력 예)
        participant = ["leo", "kiki", "eden"]
        completion = ["eden", "kiki"]
        return = "leo"

    print(solution(participant, completion))
"""


def solution(participant, completion):
    participant_dict = dict()

    for p in participant:
        participant_dict[p] = participant_dict.get(p, 0) + 1

    for c in completion:
        value = participant_dict[c] -1

        if value == 0:
            del participant_dict[c]
        else:
            participant_dict[c] = value

    answer = list(participant_dict.keys())

    return answer.pop()

File: Programmers/코딩테스트 고득점 Kit/해시/완주하지_못한_선수.java
// 완주하지 못한 선수
// https://programmers.co.kr/learn/courses/30/lessons/42576

/*
    입출력 예)
        participant = ["leo", "kiki", "eden"]
        completion = ["eden", "kiki"]
        return = "leo"

    print(solution(participant, completion))
*/


public String solution(String[] participant, String[] completion) {
        Map<String, Integer> map = new HashMap<>();
        
        for(String p : participant) {
        	map.put(p, map.getOrDefault(p, 0) +1);
        }
        
        for(String c : completion) {
        	if(map.get(c) == 1) {
        		map.remove(c);
        	}
        	else {
        		map.put(c, map.get(c) - 1);
        	}
        }
        return map.keySet().iterator().next();
    }
